fix: Use doctype setting to set auto-extracted file as private

marination opened the pull request:
- Use `make_attachments_public` to determine the file privacy while auto creating files from the text editor field
   <img width="1320" alt="Screenshot 2024-02-09 at 3 04 29 PM" src="https://github.com/frappe/frappe/assets/25857446/6e35ae8e-4e5b-471c-b79a-4f9dcbff0945">
  
   ![2024-02-09 3 11 57 PM](https://github.com/frappe/frappe/assets/25857446/322c6a6e-edad-49e6-af42-47aed34852ee)


- Currently, all files in the text editor field are automatically public and there is no way to control this behaviour except manually setting the file as private each time.

marination committed:
fix: Use doctype setting to set auto-extracted file as private

- Use `make_attachments_public` to determine the file privacy while auto creating files from the text editor field
- Currently, all files in the text editor field are automatically public

ankush commented on frappe/core/doctype/file/test_file.py:
```diff
@@ -717,16 +717,18 @@ def tearDown(self) -> None:
 
 class TestFileUtils(FrappeTestCase):
 	def test_extract_images_from_doc(self):
+		is_private = not frappe.db.get_value("DocType", "ToDo", "make_attachments_public")
```
```suggestion
		is_private = not frappe.get_meta("ToDo").make_attachments_public
```

ankush committed:
test: use meta for fetching settings

ankush commented:
TODO: 
- [ ] test communication crafted using this
- [x] test pdf crafted using this

marination committed:
test: Comment + Comm.n file extraction & Attachments default to public for Communication

- Note: DocType Communication's records are used for emails that have the file link embedded in its content
- fix: The files extracted from a Communication must be public so that they are visible in an email (link in email)
- Test: Check if file is created appropriately from a Comment and a Communication

marination commented:
@ankush not sure how to  handle `test pdf crafted using this`, i can't seem to get content out of pdfreader's pages. Would be nice if you could handle this one.

ankush commented:
@marination this doesn't work afaik

We use cookies from current request and pass it to wkhtml so it can read private images. However when it's done from background jobs it's not possible. 

https://github.com/frappe/frappe/blob/4e1ed2908c7bd9f5c6b724f3af0d26837ece87e6/frappe/utils/pdf.py#L179-L193

---

Possible fix that we have discussed before:
- For communication: send images as inline attachments with some sane max size limit (~5mb)
- For PDFs - send images as base64 encoded source to wkhtml so private file isn't a problem. 

Both are long pending design flaws, will have get it fixed somehow to make this work :pensive:

marination committed:
fix: Use regex in failing test

ankush commented:
update: working on base64-ing private images rn. Once we have this it's easier to port this change without breaking anything. 

https://github.com/frappe/frappe/pull/24980

ankush committed:
Merge branch 'develop' into txt-attachment-privacy

ankush commented on frappe/core/doctype/communication/communication.json:
```diff
@@ -401,7 +401,8 @@
  "idx": 1,
  "links": [],
- "modified": "2023-11-27 20:38:27.467076",
+ "make_attachments_public": 1,
```
This is not ideal :grimacing:

ankush replied:
Need some time to figure out inline attachments just like pdfs.

marination replied:
Yup struck me later. I was initially trying to keep the original behaviour.

ankush commented:
the world of email is :hankey: 


https://sendgrid.com/blog/embedding-images-emails-facts/

ankush committed:
Merge branch 'develop' into txt-attachment-privacy

barredterra commented:
How about this approach:

- All images dropped in Text Editor are stored as private files
- Images contained in an email are made available publicly, but non-enumerable (https://github.com/frappe/frappe/pull/25094) and under a non-guessable URL (say, 64 character hash)

This way we stay close to the current implementation but get fairly good security.

ankush committed:
Merge branch 'develop' into txt-attachment-privacy
